### PR TITLE
Update tlsio_openssl to send SNI on client hello

### PR DESCRIPTION
**Pull request by ewertons**
This change addresses github issue https://github.com/Azure/azure-c-shared-utility/issues/428
It is implemented by adding a call to SSL_set_tlsext_host_name (thanks @milope).

Verified through wireshark capture that SNI is now being sent:

```
Frame 39: 407 bytes on wire (3256 bits), 407 bytes captured (3256 bits) on interface 0
Ethernet II, Src: IntelCor_aa:bb:cc (dd:ee:ff:aa:bb:cc), Dst: Ubiquiti_4c:4d:f7 (aa:bb:cc:dd:ee:ff)
Internet Protocol Version 4, Src: 192.168.1.186, Dst: 52.180.177.125
Transmission Control Protocol, Src Port: 56250, Dst Port: 5671, Seq: 1, Ack: 1, Len: 341
Secure Sockets Layer
    TLSv1.2 Record Layer: Handshake Protocol: Client Hello
        Content Type: Handshake (22)
        Version: TLS 1.0 (0x0301)
        Length: 336
        Handshake Protocol: Client Hello
...
            Extension: server_name (len=44)
                Type: server_name (0)
                Length: 44
                Server Name Indication extension
                    Server Name list length: 42
                    Server Name Type: host_name (0)
                    Server Name length: 39
                    Server Name: contoso-iothub-000001.azure-devices.net
```

### CODE REVIEW

File: adapters/tlsio_openssl.c [hunk 1]
```diff
@@ -1036,6 +1037,7 @@ static int create_openssl_instance(TLS_IO_INSTANCE* tlsInstance)
                     }
 
                     tlsInstance->ssl = SSL_new(tlsInstance->ssl_context);
+
                     if (tlsInstance->ssl == NULL)
                     {
                         (void)BIO_free(tlsInstance->in_bio);
```
Comment (ewertons) [inline adapters/tlsio_openssl.c:1043]: Looks like a memory leak would happen here if SSL_new fail. Need to investigate.

Comment (danewalton) [reply adapters/tlsio_openssl.c:1043]: `out_bio`?

**Comment by danewalton** [comment]
sorry...but....any UT's for this?

**Comment by ewertons** [comment]
@danewalton yes, unfortunately that is a technical debt on the SDK, we have no unit tests for tlsio_openssl at all.

File: adapters/tlsio_openssl.c [hunk 2]
```diff
@@ -1045,6 +1047,17 @@ static int create_openssl_instance(TLS_IO_INSTANCE* tlsInstance)
                         log_ERR_get_error("Failed creating OpenSSL instance.");
                         result = MU_FAILURE;
                     }
+                    else if (SSL_set_tlsext_host_name(tlsInstance->ssl, tlsInstance->hostname) != 1)
```
[the remark this answers is not in the document]
Comment (CIPop) [reply adapters/tlsio_openssl.c:1050]: Also: worth testing this after forcing TLS1.0/1.1 if our Hub still supports those versions.

File: adapters/tlsio_openssl.c [hunk 3]
```diff
@@ -74,6 +74,7 @@ typedef struct TLS_IO_INSTANCE_TAG
     TLSIO_VERSION tls_version;
     TLS_CERTIFICATE_VALIDATION_CALLBACK tls_validation_callback;
     void* tls_validation_callback_data;
+    char* hostname;
```
[the remark this answers is not in the document]
Comment (ewertons) [reply adapters/tlsio_openssl.c:77]: Ctrl+F on "->hostname". See on line 1132.

File: adapters/tlsio_openssl.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (ewertons) [reply adapters/tlsio_openssl.c:1050]: That might be something to verify. I addressed the immediate github issue. Schannel does the SNI by default according to Azure Network.

**Pull request merged**